File: T-Bank/Algorithms-and-Data-Structures-2025/6sprint-Graph1/contest/D-Один_голодный_конь/solution.cpp
#include <iostream>
#include <vector>
#include <queue>


using namespace std;


void solution() {
    /*
    Time Complexity O(|V| + |E|)
    Memory Complexity O(|V| + |E|)
    */
    int n;
    scanf("%d", &n);
    int x1, y1, x2, y2;
    scanf("%d %d\n%d %d", &x1, &y1, &x2, &y2);
    vector<pair<int, int>> moves = {
        {-2, 1}, {-2, -1}, {2, 1}, {2, -1}, {-1, 2}, {1, 2}, {-1, -2}, {1, -2}
    };
    queue<
        pair<
            pair<int, int>,
            vector<pair<int, int>>
        >
    > q;
    q.push({{x1, y1}, {{x1, y1}}});
    vector<vector<bool>> seen (n + 1, vector<bool>(n + 1, false));
    seen[x1][y1] = true;
    while (!q.empty()) {
        pair<int, int> curPos = q.front().first;
        vector<pair<int, int>> path = q.front().second;
        q.pop();
        if (curPos.first == x2 && curPos.second == y2) {
            printf("%d\n", path.size() - 1);
            for (pair<int, int>& pos : path) printf("%d %d\n", pos.first, pos.second);
            return;
        }
        for (pair<int, int>& move : moves) {
            int nextX = curPos.first + move.first;
            int nextY = curPos.second + move.second;
            if (min(nextX, nextY) < 1 || max(nextX, nextY) > n || seen[nextX][nextY]) continue;
            seen[nextX][nextY] = true;
            vector<pair<int, int>> nextPath = path;
            nextPath.push_back({nextX, nextY});
            q.push({{nextX, nextY}, nextPath});
        }
    }
}

int main() {
    solution();
}

File: T-Bank/Algorithms-and-Data-Structures-2025/6sprint-Graph1/contest/D-Один_голодный_конь/solution.py
import sys
from collections import deque


def solution() -> None:
    """
    Time Complexity O(|V| + |E|)
    Memory Complexity O(|V| + |E|)
    """
    n: int = int(sys.stdin.readline().rstrip())
    x1, y1 = map(int, sys.stdin.readline().rstrip().split())
    x2, y2 = map(int, sys.stdin.readline().rstrip().split())
    moves: list[tuple[int, int]] = [
        (-2, -1), (-2, 1), (2, 1), (2, -1), (-1, 2), (1, 2), (-1, -2), (1, -2)
    ]
    seen: list[list[bool]] = [[False] * (n + 1) for _ in range(n + 1)]
    seen[x1][y1] = True
    q: list[tuple[tuple[int, int], list[tuple[int, int]]]] = deque()
    q.append(((x1, y1), [(x1, y1)]))
    while q:
        cur_pos, path = q.popleft()
        if cur_pos == (x2, y2):
            sys.stdout.write('{}\n'.format(len(path) - 1))
            sys.stdout.write('{}\n'.format('\n'.join('{} {}'.format(x, y) for x, y in path)))
            return
        for m_x, m_y in moves:
            next_x, next_y = cur_pos[0] + m_x, cur_pos[1] + m_y
            if min(next_x, next_y) < 1 or max(next_x, next_y) > n or seen[next_x][next_y]:
                continue
            next_path: list[tuple[int, int]] = path + [(next_x, next_y)]
            seen[next_x][next_y] = True
            q.append(((next_x, next_y), next_path))


if __name__ == '__main__':
    solution()